sor/src/main/java/com/bazaarvoice/emodb/sor/core/GracefulShutdownRegistry.java

```diff
@@ -0,0 +1,62 @@
+package com.bazaarvoice.emodb.sor.core;
+
+import com.bazaarvoice.emodb.common.dropwizard.lifecycle.LifeCycleRegistry;
+import com.google.inject.Inject;
+import io.dropwizard.lifecycle.Managed;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.Closeable;
+import java.io.IOException;
+import java.util.HashSet;
+import java.util.Set;
+
+import static java.util.Objects.requireNonNull;
+
+public class GracefulShutdownRegistry implements Managed {
+
+    private final static Logger _log = LoggerFactory.getLogger(GracefulShutdownRegistry.class);
+
+    private final Set<Closeable> _closeableWriters;
+    private final Set<Closeable> _closeableWriteDependents;
+
+    @Inject
+    public GracefulShutdownRegistry(LifeCycleRegistry lifeCycleRegistry) {
+        _closeableWriters = new HashSet<>();
+        _closeableWriteDependents = new HashSet<>();
+        requireNonNull(lifeCycleRegistry).manage(this);
+    }
+
+    @Override
+    public void start() throws Exception { }
+
+    @Override
+    public void stop() throws Exception {
```

Comment:
Not sure about the methodology here.  Should the writer be registered using `LifecycleRegistry` and managed by the DW `Environment` or should they register here?  Or should they be registered in both?  A more natural interface to me would be for them to register with `LifecycleRegistry` like everything else that should be shut down when the system terminates.  Then have this class simply wait until each one of those have been closed.

I'm thinking this approach was taken because there's an _ordering_ dependency in what gets shut down first.  If that's the case then maybe it should be managed another way, especially considering the already awkward way `registerWriter` and `registerWriteDependendent` are quasi-used to enforce that ordering.

Reply:
Yes the issue is the ordering component. The audit writer cannot shutdown until after the data writer has shutdown, and if dropwizard happens to call stop on them in the incorrect order a deadlock could occur, which is why i created a seperate registry with deterministic ordering.